service/updates/downloader.go

```diff
@@ -0,0 +1,344 @@
+package updates
+
+import (
+	"archive/zip"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	semver "github.com/hashicorp/go-version"
+
+	"github.com/safing/portmaster/base/log"
+)
+
+type Downloader struct {
+	dir       string
+	indexFile string
+	indexURLs []string
+	bundle    *Bundle
+	version   *semver.Version
+
+	httpClient http.Client
+}
+
+func CreateDownloader(index UpdateIndex) Downloader {
+	return Downloader{
+		dir:       index.DownloadDirectory,
+		indexFile: index.IndexFile,
+		indexURLs: index.IndexURLs,
+	}
+}
+
+func (d *Downloader) downloadIndexFile(ctx context.Context) (err error) {
+	// Make sure dir exists
+	_ = os.MkdirAll(d.dir, defaultDirMode)
+	for _, url := range d.indexURLs {
+		err = d.downloadIndexFileFromURL(ctx, url)
+		if err != nil {
+			log.Warningf("updates: failed while downloading index file %s", err)
+			continue
+		}
+		// Downloading was successful.
+		err = nil
+		break
+	}
+
+	if err == nil {
+		err = d.parseBundle()
+	}
+
+	return
+}
```

Comment:
I think it would be great to immediately parse the bundle from the downloaded data in memory instead of reading it again from disk.
Also, should one of the index URLs server a invalid file, it would skip it and try another index URL, adding more resilience.